Bridge/src/protocol.cpp

```diff
@@ -0,0 +1,162 @@
+#include "protocol.hpp"
+
+#include <algorithm>
+#include <iostream>
+#include <iterator>
+
+#include "debug.hpp"
+
+void RxProtocol::reset()
+{
+    this->body_.clear();
+    this->next_part_count_ = 0;
+}
+
+std::optional<mqtt::message_ptr> RxProtocol::receive_packet(const Frame packet)
+{
+    if (packet.frame_type != FrameType::Message)
+        // Not implemented
+        return { };
+
+    if (!this->next_frame_count_)
+    {
+        // This is the first packet to be received, start the frame counting
+        this->next_frame_count_ = packet.frame_count + 1;
+    } else if (packet.frame_count != this->next_frame_count_)
+    {
+        // We must have skipped a frame, discard everything
+        this->reset();
+        this->next_frame_count_ = packet.frame_count + 1;
+        debug << "Frame skipped" << std::endl;
+    } else
+    {
+        // Normal case, increment for next frame
+        (*this->next_frame_count_)++;
+    }
+
+    if (packet.part_count == 0)
```

Comment:
Another comment here to generalise what this group of if statement does would be good too.